psi4/src/psi4/libscf_solver/sad.cc

```diff
@@ -612,43 +675,111 @@ void SADGuess::form_gradient(SharedMatrix grad, SharedMatrix F, SharedMatrix D,
     Scratch2.reset();
 }
 
-void SADGuess::form_C_and_D(SharedMatrix X, SharedMatrix F, SharedMatrix C, SharedMatrix Cocc, SharedVector occ,
-                            SharedMatrix D) {
-    int norbs = X->rowdim();
+void SADGuess::form_C_and_D(SharedMatrix X, SharedMatrix F, SharedMatrix C, SharedVector E, SharedMatrix Cocc,
+                            SharedVector occ, SharedMatrix D) {
+    int nbf = X->rowdim();
     int nocc = occ->dim();
     if (nocc == 0) return;
 
     // Forms C in the AO basis for SAD Guesses
-    auto Scratch1 = std::make_shared<Matrix>("Scratch1", norbs, norbs);
-    auto Scratch2 = std::make_shared<Matrix>("Scratch2", norbs, norbs);
+    auto Scratch1 = std::make_shared<Matrix>("Scratch1", nbf, nbf);
+    auto Scratch2 = std::make_shared<Matrix>("Scratch2", nbf, nbf);
 
     // Form Fp = XFX
     Scratch1->gemm(true, false, 1.0, X, F, 0.0);
     Scratch2->gemm(false, false, 1.0, Scratch1, X, 0.0);
 
-    auto eigvals = std::make_shared<Vector>("Eigenvalue scratch", norbs);
-    Scratch2->diagonalize(Scratch1, eigvals);
+    // Diagonalize
+    Scratch2->diagonalize(Scratch1, E);
 
     // Form C = XC'
     C->gemm(false, false, 1.0, X, Scratch1, 0.0);
 
     // Copy over Cocc
     double** Coccp = Cocc->pointer();
     double** Cp = C->pointer();
-    for (int i = 0; i < norbs; i++) {
+    for (int i = 0; i < nbf; i++) {
         C_DCOPY(nocc, Cp[i], 1, Coccp[i], 1);
     }
     // Scale by occ
     for (int i = 0; i < nocc; i++) {
-        C_DSCAL(norbs, occ->get(i), &Coccp[0][i], nocc);
+        C_DSCAL(nbf, occ->get(i), &Coccp[0][i], nocc);
     }
     // Form D = Cocc*Cocc'
     D->gemm(false, true, 1.0, Cocc, Cocc, 0.0);
 
     Scratch1.reset();
     Scratch2.reset();
 }
+SharedMatrix SADGuess::huckel_guess() {
+    // Build Neutral D in AO basis (block diagonal)
+    SharedMatrix DAO;
+    // Huckel matrices
+    SharedMatrix Chu;
+    SharedVector Ehu;
+    run_atomic_calculations(DAO, Chu, Ehu);
+
+    IntegralFactory integral(basis_, basis_, basis_, basis_);
+    MatrixFactory mat;
+
+    int nbf = basis_->nbf();
+    mat.init_with(1, &nbf, &nbf);
+    OneBodyAOInt* S_ints = integral.ao_overlap();
```

Comment:
What about placing this pointer in a `std::unique_ptr` to avoid the call to delete below?

Reply:
Done.